server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,443 @@
+        if not self.file_list:
+            return (True, "Nothing to sync")
+        output = ""
+        list_of_files = os.path.join(self.get_working_dir(), str(uuid.uuid4()))
+        open(list_of_files, 'w').write("\n".join(sorted(self.file_list)))
+
+        # copy files here, not symlinks
+        (is_successful, this_output) = self.remote_mkdir(self.dest_directory)
+        if not is_successful:
+            _logger.error("Cannot create directory %s: %s" % (self.dest_directory, this_output))
+            return (is_successful, this_output)
+        output += this_output
+        rsync_args = self.make_rsync_args(list_of_files, self.src_directory,
+                                          self.dest_directory, self.exclude)
+        (is_successful, this_output) = self.call(rsync_args)
+        _logger.info(this_output)
+        if not is_successful:
+            _logger.error(this_output)
+            return (is_successful, this_output)
+        output += this_output
+        return (is_successful, output)
+
+    def process_main(self):
+        (successful, output) = self.rsync()
+        if not successful:
+            raise PulpCodedException(message=output)
+
+
+class Publisher(PublishStep):
+    """
+    RSync publisher class that provides the common code for publishing to remote server. Each
+    plugin needs to extend this class and add necesary logic to the _add_necesary_steps method.
+
+    :ivar symlink_list: list of symlinks to rsync
+    :ivar content_unit_file_list: list of content units to rsync
+    :ivar symlink_src: path to directory containing all symlinks
+    """
+    def __init__(self, repo, publish_conduit, config, distributor_type):
+        """
+        :param repo: Pulp managed Yum repository
+        :type  repo: pulp.plugins.model.Repository
+        :param publish_conduit: Conduit providing access to relative Pulp functionality
+        :type  publish_conduit: pulp.plugins.conduits.repo_publish.RepoPublishConduit
+        :param config: Pulp configuration for the distributor
+        :type  config: pulp.plugins.config.PluginCallConfiguration
+        :param distributor_type: The type of the distributor that is being published
+        :type distributor_type: str
+        """
+
+        super(Publisher, self).__init__("Repository publish", repo,
+                                        publish_conduit, config,
+                                        distributor_type=distributor_type)
+
+        distributor = Distributor.objects.get_or_404(repo_id=self.repo.id,
+                                                     distributor_id=publish_conduit.distributor_id)
+        self.last_published = distributor["last_publish"]
+        self.last_deleted = repo.last_unit_removed
+        self.repo = repo
+        self.predistributor = self._get_predistributor()
+
+        if self.last_published:
+            string_date = dateutils.format_iso8601_datetime(self.last_published)
+        elif self.last_published:
```

Comment:
This branch will never be taken because if `self.last_published` is `True` then the earlier branch will always be taken and this being an `elif` will never run.